### PR TITLE
Send/verify signature of ssl certificate for snode 2 snode

### PR DESCRIPTION
**Pull request by sachaaaaa**

### CODE REVIEW

**Comment by BeaudanBrown** [comment]
Seems good to me after conflicts resolved
Not sure how to test

File: httpserver/https_client.cpp [hunk 1]
```diff
@@ -28,23 +31,49 @@ void make_https_request(boost::asio::io_context& ioc,
         return;
     }
 
+    const std::string sn_pubkey_b32z =
+        std::regex_replace(sn_address, std::regex("\\.snode"), "");
```
Comment (msgmaxim) [inline httpserver/https_client.cpp]: We now have pubkey (without ".snode") in `sn_record_t`. You can get it by calling `sn_record_t::pub_key`.

File: httpserver/https_client.cpp [hunk 2]
```diff
@@ -31,20 +34,43 @@ void make_https_request(boost::asio::io_context& ioc,
...
     auto session = std::make_shared<HttpsClientSession>(
-        ioc, ctx, std::move(resolve_results), req, std::move(cb));
+        ioc, ctx, std::move(resolve_results), req, std::move(cb),
+        sn_pubkey_b32z);
 
     session->start();
 }
 
+static std::string x509_to_string(X509* x509) {
+    BIO* bio_out = BIO_new(BIO_s_mem());
```
Comment (msgmaxim) [inline httpserver/https_client.cpp]: may return NULL/nullptr

File: httpserver/https_client.cpp [hunk 3]
```diff
@@ -31,20 +34,43 @@ void make_https_request(boost::asio::io_context& ioc,
...
+    PEM_write_bio_X509(bio_out, x509);
```
Comment (msgmaxim) [inline httpserver/https_client.cpp]:
```
if (!PEM_write_bio_X509(...)) {
     /* Error */
 }
```

File: httpserver/https_client.cpp [hunk 4]
```diff
@@ -31,20 +34,43 @@ void make_https_request(boost::asio::io_context& ioc,
...
+    BUF_MEM* bio_buf;
+    BIO_get_mem_ptr(bio_out, &bio_buf);
+    std::string pem = std::string(bio_buf->data, bio_buf->length);
+    BIO_free(bio_out);
```
Comment (msgmaxim) [inline httpserver/https_client.cpp]: Will return 0 on failure

File: httpserver/CMakeLists.txt [hunk 5]
```diff
@@ -26,6 +26,7 @@ set(SRC_FILES
     rate_limiter.cpp
     https_client.cpp
     stats.cpp
+    security.cpp
```
Comment (msgmaxim) [inline httpserver/CMakeLists.txt]: include `security.h` too?

File: httpserver/security.cpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    const auto sig = generate_signature(hash, key_pair_);
+    std::string raw_sig;
+    raw_sig.reserve(sig.c.size() + sig.r.size());
+    raw_sig.insert(raw_sig.begin(), sig.c.begin(), sig.c.end());
+    raw_sig.insert(raw_sig.end(), sig.r.begin(), sig.r.end());
+    return boost::beast::detail::base64_encode(raw_sig);
+}
+
+void Security::generate_cert_signature() {
+    std::ifstream file("./cert.pem");
```
Comment (msgmaxim) [inline httpserver/security.cpp]: Need to check if the file is open (like in `lokid_key.cpp`)! Also we now can specify where the certificate file is located, so should be using the same path as in `load_server_certificate`.

File: httpserver/https_client.cpp (hunk 3 above)
Comment (sachaaaaa) [reply httpserver/https_client.cpp]: Will do

File: httpserver/https_client.cpp (hunk 2 above)
Comment (sachaaaaa) [reply httpserver/https_client.cpp]: Good point!

File: httpserver/https_client.cpp (hunk 4 above)
Comment (sachaaaaa) [reply httpserver/https_client.cpp]: Not sure if we can do anything if the free operation failed...

Comment (sachaaaaa) [reply httpserver/https_client.cpp]: Except log it?

File: httpserver/https_client.cpp [hunk 7]
```diff
@@ -84,6 +110,19 @@ void HttpsClientSession::on_write(error_code ec, size_t bytes_transferred) {
                                std::placeholders::_1, std::placeholders::_2));
 }
 
+bool HttpsClientSession::verify_signature() {
+    const auto it = res_.find(LOKI_SNODE_SIGNATURE_HEADER);
+    if (it == res_.end()) {
+        BOOST_LOG_TRIVIAL(warning)
```
Comment (msgmaxim) [inline httpserver/https_client.cpp]: We are using `LOKI_LOG` now. (the same below)

File: httpserver/https_client.cpp (hunk 4 above)
Comment (msgmaxim) [reply httpserver/https_client.cpp]: print a message and `abort()` or something?

File: httpserver/CMakeLists.txt (hunk 5 above)
Comment (sachaaaaa) [reply httpserver/CMakeLists.txt]: Yes!

File: httpserver/security.cpp (hunk 6 above)
Comment (sachaaaaa) [reply httpserver/security.cpp]: Roger that.

File: httpserver/https_client.cpp (hunk 7 above)
Comment (sachaaaaa) [reply httpserver/https_client.cpp]: Oops missed that. Thanks.

**Pull request merged**